docs/api/asyncio_con.rst

```diff
@@ -550,3 +571,154 @@ Connection Pools
     .. py:attribute:: free_size
 
         Number of available connections in the pool.
+
+
+.. _edgedb-python-asyncio-api-transaction:
+
+Transactions
+============
+
+The most robust way to make transaction is using a retry loop:
+
+.. code-block:: python
+
+    async for tx in pool.retry():
+        async with tx:
+            await tx.execute("INSERT User { name := 'Don' }")
+
+Note executing queries on ``tx`` object rather than original
+connection or connection pool.
+
+This implies that:
+
+1. Transaction is executed atomically
+2. If transaction is failed for any of the number of transient errors (i.e.
+   network failure or concurrent update error) transaction is retried
+3. If any other exception occurs, transaction is rolled back and not
+   retried.
+
+One non-obvious implication of the (3) is that whole block is retried
+including non-database statements, so for example:
+
+.. code-block:: python
+
+    async for tx in pool.retry():
+        async with tx:
+            user = await tx.fetch_one(
+                "SELECT User { email } FILTER .login = <str>$login",
+                login=login,
+            )
+            data = await httpclient.get(
+                'https://service.local/email_info',
+                params=dict(email=user.email),
+            )
+            user = await tx.fetch_one('''
+                    UPDATE User FILTER .login = <str>$login
+                    SET { email_info := <json>$data}
+                ''',
+                login=login,
+                data=data,
+            )
+
+Here execution of the http request be retried too. The core of the issue
+is that whenever transaction is interrupted user might have changed the
+email, so we have to redo all the work done.
```

Comment:
```suggestion
In the above example, the execution of the HTTP request would be
retried too.
```

I don't understand the "The core of the issue..." sentence. How can the user change the email while the code is iterating with `async for` here?

Reply:
> I don't understand the "The core of the issue..." sentence. How can the user change the email while the code is iterating with async for here?

In parallel/concurrent request for example (or a background task or whatever).

I think we should mention somehow that we retry the whole transaction not because we're lazy, but because this is how transactions must be handled. Maybe we should pick a better example, where concurrent things are more expected to happen?

Reply:
Maybe rephrase to:

"In the above example, the execution of the HTTP request would be
retried too to guarantee that the DB state is synchronized with
the remote service state."

?

Reply:
Well, I this case remote service looks like readonly, so nothing synchronized. But generally I wanted to explain why we are retrying the whole transaction block rather than individual queries. Here it's a good idea to cache the value retrieved from server and only check the cache on the second retry, so actual request isn't something that has to be done.

Reply:
Updated the first sentence and edited the text a bit: https://github.com/edgedb/edgedb-python/pull/150/commits/fc7dc2aa3e610a1f0d2fa1bf9ba88e4a4ce870b4